File: quanttide-hr-toolkit-main/packages/fastapi/src/fastapi_quanttide_hr/models/mail_message.py
from datetime import datetime

from sqlalchemy import DateTime, ForeignKey, Integer, String, Text, UniqueConstraint, func
from sqlalchemy.orm import Mapped, mapped_column

from fastapi_quanttide_hr.database import Base


class MailMessage(Base):
    __tablename__ = "mail_messages"
    __table_args__ = (UniqueConstraint("message_id", name="uq_mail_messages_message_id"),)

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    source_queue_item_id: Mapped[int | None] = mapped_column(
        Integer, ForeignKey("pending_queue.id"), nullable=True, index=True
    )
    candidate_id: Mapped[int | None] = mapped_column(
        Integer, ForeignKey("candidates.id"), nullable=True, index=True
    )
    application_id: Mapped[int | None] = mapped_column(
        Integer, ForeignKey("applications.id"), nullable=True, index=True
    )
    message_id: Mapped[str | None] = mapped_column(
        String(255), nullable=True
    )
    platform_message_id: Mapped[str | None] = mapped_column(String(255), nullable=True)

    sender_email: Mapped[str] = mapped_column(String(255))
    recipient_email: Mapped[str | None] = mapped_column(String(255), nullable=True)
    subject: Mapped[str] = mapped_column(String(500))
    body: Mapped[str | None] = mapped_column(Text, nullable=True)
    body_text: Mapped[str | None] = mapped_column(Text, nullable=True)
    attachments_json: Mapped[str | None] = mapped_column(Text, nullable=True)

    stage_snapshot: Mapped[str | None] = mapped_column(String(50), nullable=True)
    direction: Mapped[str] = mapped_column(String(20), default="inbound")

    send_status: Mapped[str | None] = mapped_column(String(20), nullable=True)
    lease_id: Mapped[str | None] = mapped_column(String(100), nullable=True)
    leased_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    retry_count: Mapped[int] = mapped_column(Integer, default=0)
    last_retry_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    failure_reason: Mapped[str | None] = mapped_column(Text, nullable=True)
    sent_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)

    occurred_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now())
    created_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now())

File: src/hr-kanban/lib/models/mail_message.dart
class MailMessage {
  final int id;
  final int? candidateId;
  final int? applicationId;
  final String? messageId;
  final String? senderEmail;
  final String? recipientEmail;
  final String subject;
  final String? body;
  final String? bodyText;
  final String? attachmentsJson;
  final String? stageSnapshot;
  final String direction;
  final String sendStatus;
  final String occurredAt;

  MailMessage({
    required this.id,
    this.candidateId,
    this.applicationId,
    this.messageId,
    this.senderEmail,
    this.recipientEmail,
    required this.subject,
    this.body,
    this.bodyText,
    this.attachmentsJson,
    this.stageSnapshot,
    required this.direction,
    this.sendStatus = '',
    required this.occurredAt,
  });

  factory MailMessage.fromJson(Map<String, dynamic> json) {
    return MailMessage(
      id: json['id'] ?? 0,
      candidateId: json['candidate_id'],
      applicationId: json['application_id'],
      messageId: json['message_id'],
      senderEmail: json['sender_email'],
      recipientEmail: json['recipient_email'],
      subject: json['subject'] ?? '',
      body: json['body'],
      bodyText: json['body_text'],
      attachmentsJson: json['attachments_json'],
      stageSnapshot: json['stage_snapshot'],
      direction: json['direction'] ?? 'inbound',
      sendStatus: json['send_status'] ?? '',
      occurredAt: json['occurred_at'] ?? '',
    );
  }
}

class TimelineItem {
  final String type;
  final String timestamp;
  final String description;
  final Map<String, dynamic>? detail;

  TimelineItem({
    required this.type,
    required this.timestamp,
    required this.description,
    this.detail,
  });

  factory TimelineItem.fromJson(Map<String, dynamic> json) {
    return TimelineItem(
      type: json['type'] ?? '',
      timestamp: json['timestamp'] ?? '',
      description: json['description'] ?? '',
      detail: json['detail'] as Map<String, dynamic>?,
    );
  }
}
